presto-main/src/main/java/com/facebook/presto/execution/buffer/PageSplitterUtil.java

```diff
@@ -26,22 +26,32 @@ private PageSplitterUtil() {}
 
     public static List<Page> splitPage(Page page, long maxPageSizeInBytes)
     {
+        return splitPage(page, maxPageSizeInBytes, Long.MAX_VALUE);
+    }
+
+    private static List<Page> splitPage(Page page, long maxPageSizeInBytes, long previousPageSize)
+    {
         checkArgument(page.getPositionCount() > 0, "page is empty");
         checkArgument(maxPageSizeInBytes > 0, "maxPageSizeInBytes must be > 0");
 
-        if (page.getSizeInBytes() <= maxPageSizeInBytes || page.getPositionCount() == 1) {
+        // for Pages with certain types of Blocks (e.g., RLE blocks) the size in bytes may remain constant
+        // through the recursive calls, which causes the recursion to only terminate when page.getPositionCount() == 1
+        // and create potentially a large number of Page's of size 1. So we check here that
+        // if the size of the page doesn't improve from the previous call we terminate the recursion.
+        if (page.getSizeInBytes() == previousPageSize || page.getSizeInBytes() <= maxPageSizeInBytes || page.getPositionCount() == 1) {
```

Comment:
My question resolves around the exact `==` check.  If there are non-rle columns, I would expect that each time the page gets a little bit smaller.  Even one byte smaller will continue the splitting.  Also, if the page grows in size we will continue to split, which we may want because some blocks expand on first split, and then shrink on subsequent splits.